File: NEsper.Tests/view/ext/TestSortWindowView.cs
///////////////////////////////////////////////////////////////////////////////////////
// Copyright (C) 2007 Esper Team. All rights reserved.                                /
// http://esper.codehaus.org                                                          /
// ---------------------------------------------------------------------------------- /
// The software in this package is published under the terms of the GPL license       /
// a copy of which has been included with this distribution in the license.txt file.  /
///////////////////////////////////////////////////////////////////////////////////////

using System;

using NUnit.Framework;

using net.esper.events;
using net.esper.support.bean;
using net.esper.support.events;
using net.esper.support.util;
using net.esper.support.view;

namespace net.esper.view.ext
{
	[TestFixture]
	public class TestSortWindowView
	{
	    private SortWindowView myView;
	    private SupportBeanClassView childView;

	    [SetUp]
	    public void SetUp()
	    {
	        // Set up length window view and a test child view
	        myView = new SortWindowView(null, new String[]{"volume"}, new Boolean[] {false}, 5, null);
	        childView = new SupportBeanClassView(typeof(SupportMarketDataBean));
	        myView.AddView(childView);
	    }

	    [Test]
	    public void testViewOneProperty()
	    {
	        // Set up a feed for the view under test - the depth is 10 events so bean[10] will cause bean[0] to go old
	        SupportStreamImpl stream = new SupportStreamImpl(typeof(SupportMarketDataBean), 10);
	        stream.AddView(myView);

	        EventBean[] bean = new EventBean[12];

	        bean[0] = MakeBean(1000);
	        stream.Insert(bean[0]);
	        SupportViewDataChecker.CheckOldData(childView, null);
	        SupportViewDataChecker.CheckNewData(childView, new EventBean[] { bean[0] });
	        ArrayAssertionUtil.AreEqualExactOrder(myView.GetEnumerator(), new EventBean[] { bean[0] });

	        bean[1] = MakeBean(800);
	        bean[2] = MakeBean(1200);
	        stream.Insert(new EventBean[] { bean[1], bean[2] });
	        SupportViewDataChecker.CheckOldData(childView, null);
	        SupportViewDataChecker.CheckNewData(childView, new EventBean[] { bean[1], bean[2] });
	        ArrayAssertionUtil.AreEqualExactOrder(myView.GetEnumerator(), new EventBean[] { bean[1], bean[0], bean[2] });

	        bean[3] = MakeBean(1200);
	        bean[4] = MakeBean(1000);
	        bean[5] = MakeBean(1400);
	        bean[6] = MakeBean(1100);
	        stream.Insert(new EventBean[] { bean[3], bean[4], bean[5], bean[6] });
	        SupportViewDataChecker.CheckOldData(childView, new EventBean[] { bean[5], bean[2] });
	        SupportViewDataChecker.CheckNewData(childView, new EventBean[] { bean[3], bean[4], bean[5], bean[6] });
	        ArrayAssertionUtil.AreEqualExactOrder(myView.GetEnumerator(), new EventBean[] { bean[1], bean[4], bean[0], bean[6], bean[3] });

	        bean[7] = MakeBean(800);
	        bean[8] = MakeBean(700);
	        bean[9] = MakeBean(1200);
	        stream.Insert(new EventBean[] { bean[7], bean[8], bean[9] });
	        SupportViewDataChecker.CheckOldData(childView, new EventBean[] { bean[3], bean[9], bean[6] });
	        SupportViewDataChecker.CheckNewData(childView, new EventBean[] { bean[7], bean[8], bean[9] });
	        ArrayAssertionUtil.AreEqualExactOrder(myView.GetEnumerator(), new EventBean[] { bean[8], bean[7], bean[1], bean[4], bean[0] });

	        bean[10] = MakeBean(1050);
	        stream.Insert(new EventBean[] { bean[10] });       // Thus bean[0] will be old data !
	        SupportViewDataChecker.CheckOldData(childView, new EventBean[] { bean[0] });
	        SupportViewDataChecker.CheckNewData(childView, new EventBean[] { bean[10] });
	        ArrayAssertionUtil.AreEqualExactOrder(myView.GetEnumerator(), new EventBean[] { bean[8], bean[7], bean[1], bean[4], bean[10] });

	        bean[11] = MakeBean(2000);
	        stream.Insert(new EventBean[] { bean[11] });       // Thus bean[1] will be old data !
	        SupportViewDataChecker.CheckOldData(childView, new EventBean[] { bean[1] });
	        SupportViewDataChecker.CheckNewData(childView, new EventBean[] { bean[11] });
	        ArrayAssertionUtil.AreEqualExactOrder(myView.GetEnumerator(), new EventBean[] { bean[8], bean[7], bean[4], bean[10], bean[11] });
	    }

	    [Test]
	    public void testViewTwoProperties()
	    {
	    	// Set up a sort windows that sorts on two properties
	    	myView = new SortWindowView(null, new String[]{"volume", "price"}, new Boolean[] {false, true}, 5, null);
	        childView = new SupportBeanClassView(typeof(SupportMarketDataBean));
	        myView.AddView(childView);

	        // Set up a feed for the view under test - the depth is 10 events so bean[10] will cause bean[0] to go old
	        SupportStreamImpl stream = new SupportStreamImpl(typeof(SupportMarketDataBean), 10);
	        stream.AddView(myView);

	        EventBean[] bean = new EventBean[12];

	        bean[0] = MakeBean(20d, 1000);
	        stream.Insert(bean[0]);
	        SupportViewDataChecker.CheckOldData(childView, null);
	        SupportViewDataChecker.CheckNewData(childView, new EventBean[] { bean[0] });
	        ArrayAssertionUtil.AreEqualExactOrder(myView.GetEnumerator(), new EventBean[] { bean[0] });

	        bean[1] = MakeBean(19d, 800);
	        bean[2] = MakeBean(18d, 1200);
	        stream.Insert(new EventBean[] { bean[1], bean[2] });
	        SupportViewDataChecker.CheckOldData(childView, null);
	        SupportViewDataChecker.CheckNewData(childView, new EventBean[] { bean[1], bean[2] });
	        ArrayAssertionUtil.AreEqualExactOrder(myView.GetEnumerator(), new EventBean[] { bean[1], bean[0], bean[2] });

	        bean[3] = MakeBean(17d, 1200);
	        bean[4] = MakeBean(16d, 1000);
	        bean[5] = MakeBean(15d, 1400);
	        bean[6] = MakeBean(14d, 1100);
	        stream.Insert(new EventBean[] { bean[3], bean[4], bean[5], bean[6] });
	        SupportViewDataChecker.CheckOldData(childView, new EventBean[] { bean[5], bean[3] });
	        SupportViewDataChecker.CheckNewData(childView, new EventBean[] { bean[3], bean[4], bean[5], bean[6] });
	        ArrayAssertionUtil.AreEqualExactOrder(myView.GetEnumerator(), new EventBean[] { bean[1], bean[0], bean[4], bean[6], bean[2] });

	        bean[7] = MakeBean(13d, 800);
	        bean[8] = MakeBean(12d, 700);
	        bean[9] = MakeBean(11d, 1200);
	        stream.Insert(new EventBean[] { bean[7], bean[8], bean[9] });
	        SupportViewDataChecker.CheckOldData(childView, new EventBean[] { bean[9], bean[2], bean[6] });
	        SupportViewDataChecker.CheckNewData(childView, new EventBean[] { bean[7], bean[8], bean[9] });
	        ArrayAssertionUtil.AreEqualExactOrder(myView.GetEnumerator(), new EventBean[] { bean[8], bean[1], bean[7], bean[0], bean[4] });

	        bean[10] = MakeBean(10d, 1050);
	        stream.Insert(new EventBean[] { bean[10] });       // Thus bean[0] will be old data !
	        SupportViewDataChecker.CheckOldData(childView, new EventBean[] { bean[0] });
	        SupportViewDataChecker.CheckNewData(childView, new EventBean[] { bean[10] });
	        ArrayAssertionUtil.AreEqualExactOrder(myView.GetEnumerator(), new EventBean[] { bean[8], bean[1], bean[7], bean[4], bean[10] });

	        bean[11] = MakeBean(2000);
	        stream.Insert(new EventBean[] { bean[11] });       // Thus bean[1] will be old data !
	        SupportViewDataChecker.CheckOldData(childView, new EventBean[] { bean[1] });
	        SupportViewDataChecker.CheckNewData(childView, new EventBean[] { bean[11] });
	        ArrayAssertionUtil.AreEqualExactOrder(myView.GetEnumerator(), new EventBean[] { bean[8], bean[7], bean[4], bean[10], bean[11] });
	    }

	    private EventBean MakeBean(long volume)
	    {
	        SupportMarketDataBean bean = new SupportMarketDataBean("CSCO.O", 0, volume, "");
	        return SupportEventBeanFactory.CreateObject(bean);
	    }

	    private EventBean MakeBean(double price, long volume)
	    {
	        SupportMarketDataBean bean = new SupportMarketDataBean("CSCO.O", price, volume, "");
	        return SupportEventBeanFactory.CreateObject(bean);
	    }
	}
} // End of namespace

File: esper/src/test/java/com/espertech/esper/view/ext/TestSortWindowView.java
package com.espertech.esper.view.ext;

import junit.framework.TestCase;
import com.espertech.esper.client.EventBean;
import com.espertech.esper.support.bean.SupportMarketDataBean;
import com.espertech.esper.support.event.SupportEventBeanFactory;
import com.espertech.esper.support.util.ArrayAssertionUtil;
import com.espertech.esper.support.view.SupportBeanClassView;
import com.espertech.esper.support.view.SupportStreamImpl;
import com.espertech.esper.support.view.SupportViewDataChecker;
import com.espertech.esper.support.epl.SupportExprNodeFactory;

public class TestSortWindowView extends TestCase
{
    private SortWindowView myView;
    private SupportBeanClassView childView;

    public void setUp() throws Exception
    {
        // Set up length window view and a test child view
        myView = new SortWindowView(null, SupportExprNodeFactory.makeIdentNodesMD("volume"), new boolean[] {false}, 5, null, false);
        childView = new SupportBeanClassView(SupportMarketDataBean.class);
        myView.addView(childView);
    }
    
    public void testViewOneProperty()
    {
        // Set up a feed for the view under test - the depth is 10 events so bean[10] will cause bean[0] to go old
        SupportStreamImpl stream = new SupportStreamImpl(SupportMarketDataBean.class, 10);
        stream.addView(myView);

        EventBean bean[] = new EventBean[12];

        bean[0] = makeBean(1000);
        stream.insert(bean[0]);
        SupportViewDataChecker.checkOldData(childView, null);
        SupportViewDataChecker.checkNewData(childView, new EventBean[] { bean[0] });
        ArrayAssertionUtil.assertEqualsExactOrder(myView.iterator(), new EventBean[] { bean[0] });

        bean[1] = makeBean(800);
        bean[2] = makeBean(1200);
        stream.insert(new EventBean[] { bean[1], bean[2] });
        SupportViewDataChecker.checkOldData(childView, null);
        SupportViewDataChecker.checkNewData(childView, new EventBean[] { bean[1], bean[2] });
        ArrayAssertionUtil.assertEqualsExactOrder(myView.iterator(), new EventBean[] { bean[1], bean[0], bean[2] });

        bean[3] = makeBean(1200);
        bean[4] = makeBean(1000);
        bean[5] = makeBean(1400);
        bean[6] = makeBean(1100);
        stream.insert(new EventBean[] { bean[3], bean[4], bean[5], bean[6] });
        SupportViewDataChecker.checkOldData(childView, new EventBean[] { bean[5], bean[2] });
        SupportViewDataChecker.checkNewData(childView, new EventBean[] { bean[3], bean[4], bean[5], bean[6] });
        ArrayAssertionUtil.assertEqualsExactOrder(myView.iterator(), new EventBean[] { bean[1], bean[4], bean[0], bean[6], bean[3] });

        bean[7] = makeBean(800);
        bean[8] = makeBean(700);
        bean[9] = makeBean(1200);
        stream.insert(new EventBean[] { bean[7], bean[8], bean[9] });
        SupportViewDataChecker.checkOldData(childView, new EventBean[] { bean[3], bean[9], bean[6] });
        SupportViewDataChecker.checkNewData(childView, new EventBean[] { bean[7], bean[8], bean[9] });
        ArrayAssertionUtil.assertEqualsExactOrder(myView.iterator(), new EventBean[] { bean[8], bean[7], bean[1], bean[4], bean[0] });

        bean[10] = makeBean(1050);
        stream.insert(new EventBean[] { bean[10] });       // Thus bean[0] will be old data !
        SupportViewDataChecker.checkOldData(childView, new EventBean[] { bean[0] });
        SupportViewDataChecker.checkNewData(childView, new EventBean[] { bean[10] });
        ArrayAssertionUtil.assertEqualsExactOrder(myView.iterator(), new EventBean[] { bean[8], bean[7], bean[1], bean[4], bean[10] });

        bean[11] = makeBean(2000);
        stream.insert(new EventBean[] { bean[11] });       // Thus bean[1] will be old data !
        SupportViewDataChecker.checkOldData(childView, new EventBean[] { bean[1] });
        SupportViewDataChecker.checkNewData(childView, new EventBean[] { bean[11] });
        ArrayAssertionUtil.assertEqualsExactOrder(myView.iterator(), new EventBean[] { bean[8], bean[7], bean[4], bean[10], bean[11] });
    }
    
    public void testViewTwoProperties() throws Exception
    {
    	// Set up a sort windows that sorts on two properties
    	myView = new SortWindowView(null, SupportExprNodeFactory.makeIdentNodesMD("volume", "price"), new boolean[] {false, true}, 5, null, false);
        childView = new SupportBeanClassView(SupportMarketDataBean.class);
        myView.addView(childView);
        
        // Set up a feed for the view under test - the depth is 10 events so bean[10] will cause bean[0] to go old
        SupportStreamImpl stream = new SupportStreamImpl(SupportMarketDataBean.class, 10);
        stream.addView(myView);
        
        EventBean bean[] = new EventBean[12];
        
        bean[0] = makeBean(20d, 1000);
        stream.insert(bean[0]);
        SupportViewDataChecker.checkOldData(childView, null);
        SupportViewDataChecker.checkNewData(childView, new EventBean[] { bean[0] });
        ArrayAssertionUtil.assertEqualsExactOrder(myView.iterator(), new EventBean[] { bean[0] });
 
        bean[1] = makeBean(19d, 800);
        bean[2] = makeBean(18d, 1200);
        stream.insert(new EventBean[] { bean[1], bean[2] });
        SupportViewDataChecker.checkOldData(childView, null);
        SupportViewDataChecker.checkNewData(childView, new EventBean[] { bean[1], bean[2] });
        ArrayAssertionUtil.assertEqualsExactOrder(myView.iterator(), new EventBean[] { bean[1], bean[0], bean[2] });
        
        bean[3] = makeBean(17d, 1200);
        bean[4] = makeBean(16d, 1000);
        bean[5] = makeBean(15d, 1400);
        bean[6] = makeBean(14d, 1100);
        stream.insert(new EventBean[] { bean[3], bean[4], bean[5], bean[6] });
        SupportViewDataChecker.checkOldData(childView, new EventBean[] { bean[5], bean[3] });
        SupportViewDataChecker.checkNewData(childView, new EventBean[] { bean[3], bean[4], bean[5], bean[6] });
        ArrayAssertionUtil.assertEqualsExactOrder(myView.iterator(), new EventBean[] { bean[1], bean[0], bean[4], bean[6], bean[2] });

        bean[7] = makeBean(13d, 800);
        bean[8] = makeBean(12d, 700);
        bean[9] = makeBean(11d, 1200);
        stream.insert(new EventBean[] { bean[7], bean[8], bean[9] });
        SupportViewDataChecker.checkOldData(childView, new EventBean[] { bean[9], bean[2], bean[6] });
        SupportViewDataChecker.checkNewData(childView, new EventBean[] { bean[7], bean[8], bean[9] });
        ArrayAssertionUtil.assertEqualsExactOrder(myView.iterator(), new EventBean[] { bean[8], bean[1], bean[7], bean[0], bean[4] });

        bean[10] = makeBean(10d, 1050);
        stream.insert(new EventBean[] { bean[10] });       // Thus bean[0] will be old data !
        SupportViewDataChecker.checkOldData(childView, new EventBean[] { bean[0] });
        SupportViewDataChecker.checkNewData(childView, new EventBean[] { bean[10] });
        ArrayAssertionUtil.assertEqualsExactOrder(myView.iterator(), new EventBean[] { bean[8], bean[1], bean[7], bean[4], bean[10] });

        bean[11] = makeBean(2000);
        stream.insert(new EventBean[] { bean[11] });       // Thus bean[1] will be old data !
        SupportViewDataChecker.checkOldData(childView, new EventBean[] { bean[1] });
        SupportViewDataChecker.checkNewData(childView, new EventBean[] { bean[11] });
        ArrayAssertionUtil.assertEqualsExactOrder(myView.iterator(), new EventBean[] { bean[8], bean[7], bean[4], bean[10], bean[11] });
    }

    private EventBean makeBean(long volume)
    {
        SupportMarketDataBean bean = new SupportMarketDataBean("CSCO.O", 0, volume, "");
        return SupportEventBeanFactory.createObject(bean);
    }
    
    private EventBean makeBean(double price, long volume)
    {
        SupportMarketDataBean bean = new SupportMarketDataBean("CSCO.O", price, volume, "");
        return SupportEventBeanFactory.createObject(bean);
    }
}
